kura/org.eclipse.kura.wire.ai.component.provider/src/main/java/org/eclipse/kura/wire/ai/component/provider/AIComponent.java

```diff
@@ -0,0 +1,209 @@
+import org.eclipse.kura.configuration.ConfigurableComponent;
+import org.eclipse.kura.wire.WireComponent;
+import org.eclipse.kura.wire.WireEmitter;
+import org.eclipse.kura.wire.WireEnvelope;
+import org.eclipse.kura.wire.WireHelperService;
+import org.eclipse.kura.wire.WireReceiver;
+import org.eclipse.kura.wire.WireRecord;
+import org.eclipse.kura.wire.WireSupport;
+import org.osgi.framework.ServiceReference;
+import org.osgi.service.component.ComponentContext;
+import org.osgi.service.component.ComponentException;
+import org.osgi.service.wireadmin.Wire;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class AIComponent implements WireEmitter, WireReceiver, ConfigurableComponent {
+
+    private static final Logger logger = LoggerFactory.getLogger(AIComponent.class);
+
+    private volatile WireHelperService wireHelperService;
+    private WireSupport wireSupport;
+
+    private AIComponentOptions options;
+    private InferenceEngineService inferenceEngineService;
+
+    private Optional<ModelInfo> infoPre;
+    private Optional<ModelInfo> infoInfer;
+    private Optional<ModelInfo> infoPost;
+
+    public void bindWireHelperService(final WireHelperService wireHelperService) {
+        if (this.wireHelperService == null) {
+            this.wireHelperService = wireHelperService;
+        }
+    }
+
+    public void unbindWireHelperService(final WireHelperService wireHelperService) {
+        if (this.wireHelperService == wireHelperService) {
+            this.wireHelperService = null;
+        }
+    }
+
+    public void bindInferenceEngineService(final InferenceEngineService inferenceEngineService) {
+        this.inferenceEngineService = inferenceEngineService;
+    }
+
+    public void unbindInferenceEngineService(final InferenceEngineService inferenceEngineService) {
+        if (this.inferenceEngineService == inferenceEngineService) {
+            this.inferenceEngineService = inferenceEngineService;
+        }
+    }
+
+    public void activate(final ComponentContext componentContext, final Map<String, Object> properties)
+            throws ComponentException {
+        logger.info("Activating AIComponent...");
+
+        this.wireSupport = this.wireHelperService.newWireSupport(this,
+                (ServiceReference<WireComponent>) componentContext.getServiceReference());
+
+        updated(properties);
+
+        logger.info("Activating AIComponent... Done");
+    }
+
+    public void deactivate() {
+        logger.info("Deactivating AIComponent...");
+        logger.info("Deactivating AIComponent... Done.");
+    }
+
+    public synchronized void updated(final Map<String, Object> properties) {
+        logger.info("Updating AIComponent...");
+
+        this.options = new AIComponentOptions(properties);
+        this.infoPre = Optional.empty();
+        this.infoInfer = Optional.empty();
+        this.infoPost = Optional.empty();
+
+        if (this.inferenceEngineService != null) {
```

Comment:
Should we run this part in onWireReceive? otherwise if the engine is not ready when the activate/update is called, model/modelInfo loading may fail. If this is done in onWireReceive we will have the chance to retry. Of course we can cache the model infos until next update.

Reply:
I would put the check in both methods